pkg/fanal/analyzer/language/python/poetry/poetry.go

```diff
@@ -105,34 +105,75 @@ func (a poetryAnalyzer) mergePyProject(fsys fs.FS, dir string, app *types.Applic
 
 	// Identify the direct/transitive dependencies
 	for i, pkg := range app.Packages {
-		if _, ok := p[pkg.Name]; ok {
+		if _, ok := project.Tool.Poetry.Dependencies[pkg.Name]; ok {
 			app.Packages[i].Relationship = types.RelationshipDirect
 		} else {
 			app.Packages[i].Indirect = true
 			app.Packages[i].Relationship = types.RelationshipIndirect
 		}
 	}
 
+	prodDeps := getProdDeps(project, app)
+
+	app.Packages = lo.Filter(app.Packages, func(pkg types.Package, _ int) bool {
+		_, ok := prodDeps[pkg.ID]
+		return ok
+	})
+
 	return nil
 }
 
-func (a poetryAnalyzer) parsePyProject(fsys fs.FS, path string) (map[string]any, error) {
+func getProdDeps(project pyproject.PyProject, app *types.Application) map[string]struct{} {
+	packages := lo.SliceToMap(app.Packages, func(pkg types.Package) (string, types.Package) {
+		return pkg.ID, pkg
+	})
+
+	visited := make(map[string]struct{})
+	deps := project.Tool.Poetry.Dependencies
+
+	for group, groupDeps := range project.Tool.Poetry.Groups {
+		if group == "dev" {
+			continue
+		}
+		deps = lo.Assign(deps, groupDeps.Dependencies)
+	}
+
+	for _, pkg := range packages {
+		if _, prodDep := deps[pkg.Name]; !prodDep {
+			continue
+		}
+		walkPackageDeps(pkg.ID, packages, visited)
+	}
+	return visited
+}
+
+func walkPackageDeps(pkgID string, packages map[string]types.Package, visited map[string]struct{}) {
+	if _, ok := visited[pkgID]; ok {
+		return
+	}
+	visited[pkgID] = struct{}{}
+	pkg, exists := packages[pkgID]
+	if !exists {
+		return
+	}
```

Comment:
Do we need to check if a package with `pkgID` exists before adding it to `visited`?
IIUC we check for this case in the parser:
https://github.com/aquasecurity/trivy/blob/faa9d92cfeb8d924deda2dac583b6c97099c08d9/pkg/dependency/parser/python/poetry/parse.go#L109-L112
but I would still prefer to leave the "correct" order of actions

wdyt?

Reply:
Should we be strict when checking? A package can only be missing in 1 case if `pyproject.toml` is manually updated and a new dependency is added and no dependency locking is performed, but the `lockfile` is still considered valid.

Reply:
>A package can only be missing in 1 case if pyproject.toml is manually updated and a new dependency is added and no dependency locking is performed, but the lockfile is still considered valid.

let's stay this check for this case. 
sometimes users forget to update the lock file after updating the "main" file
this check will remove ambiguities

Reply:
I was a little mistaken. With recursive traversal, the packet is guaranteed to exist. It can only be missing [here](https://github.com/aquasecurity/trivy/pull/8106/files#diff-299990016b1ff5e9d6606f989344ae472ca4f05f4d74a4f577478d3800586f6bR136-R138) for two reasons: the user removed the package manually from `pyproject.toml` or the package is not a direct dependency.

Reply:
okay, then we can remove this check.